src/query.rs

```diff
@@ -8,113 +8,182 @@ use http::{
+                        snowflake_type: row_type.data_type,
+                        nullable: row_type.nullable,
+                    },
+                )
+            })
+            .collect::<HashMap<_, _>>();
+        let column_types = Arc::new(column_types);
+
+        let chunk_headers = response.data.chunk_headers.unwrap_or_default();
+        let chunk_headers: HeaderMap = HeaderMap::try_from(&chunk_headers)?;
+
+        Ok(Self {
+            http,
+            qrmk,
+            chunks,
+            chunk_headers,
+            column_types,
+            row_set: Some(row_set),
+        })
     }
 
-    if !response.success {
-        return Err(Error::Communication(response.message.unwrap_or_default()));
+    /// Check if there are no more rows to fetch
+    pub fn eof(&self) -> bool {
+        self.row_set.is_none() && self.chunks.is_empty()
     }
 
-    if let Some(format) = response.data.query_result_format {
-        if format != "json" {
-            return Err(Error::UnsupportedFormat(format.clone()));
+    /// Fetch a single chunk
+    pub async fn fetch_chunk(&mut self) -> Result<Vec<SnowflakeRow>> {
+        if let Some(row_set) = self.row_set.take() {
+            let rows = row_set.into_iter().map(|r| self.convert_row(r)).collect();
+            return Ok(rows);
         }
-    }
 
-    let http = http.clone();
-    let qrmk = response.data.qrmk.unwrap_or_default();
-    let chunks = response.data.chunks.unwrap_or_default();
-    let row_types = response.data.row_types.ok_or_else(|| {
-        Error::UnsupportedFormat("the response doesn't contain 'rowtype'".to_string())
-    })?;
-    let mut row_set = response.data.row_set.ok_or_else(|| {
-        Error::UnsupportedFormat("the response doesn't contain 'rowset'".to_string())
-    })?;
-
-    let chunk_headers = response.data.chunk_headers.unwrap_or_default();
-    let chunk_headers: HeaderMap = HeaderMap::try_from(&chunk_headers)?;
-
-    let mut handles = Vec::with_capacity(chunks.len());
-    for chunk in chunks {
-        let http = http.clone();
-        let chunk_headers = chunk_headers.clone();
-        let qrmk = qrmk.clone();
-        handles.push(tokio::spawn(async move {
-            download_chunk(http, chunk.url, chunk_headers, qrmk).await
-        }));
+        let http = self.http.clone();
+        let chunk_headers = self.chunk_headers.clone();
+        let qrmk = self.qrmk.clone();
+        let chunk = self
+            .chunks
+            .pop()
+            .ok_or(Error::ChunkDownload("no more chunks to fetch".to_owned()))?;
+        let rows = download_chunk(http, chunk.url, chunk_headers, qrmk).await?;
+        let rows = rows.into_iter().map(|r| self.convert_row(r)).collect();
+        Ok(rows)
     }
 
-    for fut in handles {
-        let result = fut.await?;
-        let rows = result?;
-        row_set.extend(rows);
+    /// Fetch all the remaining chunks at once
+    pub async fn fetch_all(&mut self) -> Result<Vec<SnowflakeRow>> {
```

Comment:
This function returns:
- `Ok(None)` if there's no remaining chunk.
- `Ok(Some(rows))` if it successfully fetched the next chunk.
- `Err(e)` if it failed to fetch the next chunk.